connect_ferc1_to_eia.py

```diff
@@ -43,496 +43,894 @@
 from sklearn.model_selection import KFold  # , cross_val_score
 
 import pudl
+import make_plant_parts_eia
 
 logger = logging.getLogger(__name__)
 
 
-def prep_train_connections(file_path_training, plant_parts_df):
-    """
-    Get and prepare the training connections.
-
-    We have stored connections between ferc1 steam and the eia master unit
-    list. These records should be compiled of connections between these data
-    sets that are known to be correct because we will use these records to
-    train a machine learning model.
-
-    Args:
-        training_file_path (path-like): path to the CSV of training data. The
-            training data needs to have at least two columns: record_id_eia and
-            record_id_ferc1.
-        plant_parts_df (pandas.DataFrame): master unit list. generated from
-            `CompilePlantParts.generate_master_unit_list()` or
-            `get_master_unit_list_eia()`.
-
-    Returns:
-        pandas.DataFrame: training connections. A dataframe with has a
-        MultiIndex with record_id_eia and record_id_ferc1.
-    """
-    mul_cols = ['true_gran', 'appro_part_label',
-                'appro_record_id_eia', 'plant_part']
-    train_df = (
-        pd.read_csv(file_path_training,)
-        # we want to ensure that the records are associated with a
-        # "true granularity" - which is a way we filter out whether or not each
-        # record in the master unit list is actually a new/unique collection of
-        # plant parts
-        .merge(
-            plant_parts_df.reset_index()[['record_id_eia'] + mul_cols],
-            how='left', on=['record_id_eia'],
+class Inputs:
```

Comment:
It looks to me like the *Inputs* object is more or less a tool to manage the process of shaping your inputs, rather than a representation of the Input itself.

Specifically - unless I'm mistaken, you *don't* instantiate an **Inputs** object and then *pass it around* or *do things to it.*  Instead, you're using it to *produce outputs that you will use elsewhere.*

With that model, my preference would be that *InputManager*:

- Recieve and store *process configuration* at *__init__.* (Especially **clobber** and any **pudl_settings** type details.)
    - This will allow you do set and forget the setup.
- Favor a *fixed input -> output* model for data processing functions.  Rule of thumb, "this method relies on" should not exist, take an input and produce an output.
   - Benefits:
        - Each stage to be independently tested.
        - One stage won't interfere with another if repeated or altered during maintenance.
        - Future programmers will not need to keep track of side effects. Cognitive load is a major issue when you rely on side effects as part of the data process.
- Handle coordination and memoization of the complete process with something conceptually similar to a __main__ function, even if it needs another name, like InputManager.construct_dataset(*inputs).
   - This ends up doing double duty; both in coordinating all the steps and specifying exactly how all the other functions work together for the next coder.

I hope this reads more like some potentially helpful ideas than criticisms.

Reply:
this is very helpful! Thanks!

I will change the name to `InputManager`. That is much more clear.

- For the "process configuration" comment, I believe I understand this. I think I did this backwards last time around, but does the current setup conform to this?
- Same comment/question for the input -> output model comment. I removed the "this method relies on" comments. I double checked whether or not I had implemented the pattern where instead of directly using cached tables, it now runs a method which checks whether or not the table has already been cached.
- I love the last point! I added an `execute` method... not sure if that is the best name but I have seen it be used in these types of contexts.